Source/Plugin.swift

```diff
@@ -8,16 +8,24 @@ import Result
 ///     - hide and show a network activity indicator
 ///     - inject additional information into a request
 public protocol PluginType {
+    /// Called to modify a request before sending
+    func prepareRequest(_ request: URLRequest, target: TargetType) -> URLRequest
+
     /// Called immediately before a request is sent over the network (or stubbed).
     func willSendRequest(_ request: RequestType, target: TargetType)
 
     /// Called after a response has been received, but before the MoyaProvider has invoked its completion handler.
     func didReceiveResponse(_ result: Result<Moya.Response, Moya.Error>, target: TargetType)
+
+    /// Called to modify a result before completion
+    func processResponse(_ result: Result<Moya.Response, Moya.Error>, target: TargetType) -> Result<Moya.Response, Moya.Error>
 }
 
 public extension PluginType {
+    func prepareRequest(_ request: URLRequest, target: TargetType) -> URLRequest { return request }
```

Comment:
My hope was that the type information would be clear enough, but I can definitely see both sides. We can't use `RequestType` for `prepareRequest`, but it would also be possible to change `willSendRequest` to pass just the `URLRequest` of the `Alamofire.Request` to be more consistent. This would eliminate the need for `RequestType` but it would also mean that `CredentialsPlugin` would need to be modified. That might be a good thing now though so that there is consistency in where in the lifecycle requests are being modified.